test.cc

```diff
@@ -275,3 +277,192 @@ TEST(Sherlock, SubscribeToStreamViaHTTP) {
+    explicit CallBackTest(int key, double value, bool expect_fail = false) :
+        key(key), value(value), expect_fail(expect_fail) {}
+
+    void found(const KeyValueEntry& entry) {
+      called = true;
+      EXPECT_FALSE(expect_fail);
+      EXPECT_EQ(key, entry.key()());
+      EXPECT_EQ(value, entry.value_);
+    }
+    void not_found(const IntKey& key) {
+      called = true;
+      EXPECT_TRUE(expect_fail);
+      EXPECT_EQ(this->key, key());
+    }
+    void added() {
+      called = true;
+      EXPECT_FALSE(expect_fail);
+    }
+    void already_exists() {
+      called = true;
+      EXPECT_TRUE(expect_fail);
+    }
+
+    const int key;
+    const double value;
+    const bool expect_fail;
+    bool called = false;
+  };
+
+  CallBackTest cbt1(2, 0.5);
+  auto cbf1 = [&](const KeyValueEntry& entry) { cbt1.found(entry); };
+  auto cbnf1 = [&](const IntKey& key) { cbt1.not_found(key); };
+  api.AsyncGet(KVS::T_KEY(2), cbf1, cbnf1);
+  while (!cbt1.called)
+    ;
+
+  // Add two more key-value pairs.
+  api.UnsafeStream().Emplace(3, 0.33);
+  api.UnsafeStream().Emplace(4, 0.25);
+
+  while (api.EntriesSeen() < 3u) {
+    // For the purposes of this test: Spin lock to ensure that the listener/MMQ consumer got the data published.
+  }
+
+  EXPECT_EQ(0.33, api.AsyncGet(KVS::T_KEY(3)).get().value_);
+  EXPECT_EQ(0.25, api.Get(KVS::T_KEY(4)).value_);
+
+  ASSERT_THROW(api.AsyncGet(KVS::T_KEY(5)).get(), KVS::T_KEY_NOT_FOUND_EXCEPTION);
+  ASSERT_THROW(api.Get(KVS::T_KEY(6)), KVS::T_KEY_NOT_FOUND_EXCEPTION);
+  CallBackTest cbt2(7, 0.0, true);
+  auto cbf2 = [&](const KeyValueEntry& entry) { cbt2.found(entry); };
+  auto cbnf2 = [&](const IntKey& key) { cbt2.not_found(key); };
+  api.AsyncGet(KVS::T_KEY(7), cbf2, cbnf2);
+  while (!cbt2.called)
+    ;
+
+  // Add three more key-value pairs, this time via the API.
+  api.AsyncAdd(KeyValueEntry(5, 0.2)).wait();
+  api.Add(KeyValueEntry(6, 0.17));
+  CallBackTest cbt3(7, 0.76);
+  auto cba3 = [&]() { cbt3.added(); };
+  auto cbae3 = [&]() { cbt3.already_exists(); };
+  api.AsyncAdd(KeyValueEntry(7, 0.76), cba3, cbae3);
+  while (!cbt3.called)
+    ;
+
+  // Check that default policy doesn't allow overwriting on Add().
+  ASSERT_THROW(api.AsyncAdd(KeyValueEntry(5, 1.1)).get(), KVS::T_KEY_ALREADY_EXISTS_EXCEPTION);
+  ASSERT_THROW(api.Add(KeyValueEntry(6, 0.28)), KVS::T_KEY_ALREADY_EXISTS_EXCEPTION);
+  CallBackTest cbt4(7, 0.0, true);
+  auto cba4 = [&]() { cbt4.added(); };
+  auto cbae4 = [&]() { cbt4.already_exists(); };
+  api.AsyncAdd(KeyValueEntry(7, 0.0), cba4, cbae4);
+  while (!cbt4.called)
+    ;
+
+  // Thanks to eventual consistency, we don't have to wait until the above calls fully propagate.
+  // Even if the next two lines run before the entries are published into the stream,
+  // the API will maintain the consistency of its own responses from its own in-memory state.
+  EXPECT_EQ(0.20, api.AsyncGet(KVS::T_KEY(5)).get().value_);
+  EXPECT_EQ(0.17, api.Get(KVS::T_KEY(6)).value_);
```

Comment:
I'd change at least a few of these lines from `KVS::T_KEY` to the actual type, just to illustrate it.